aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move

```diff
@@ -0,0 +1,529 @@
+    struct Entry<K, V> has drop, copy, store {
+        key: K,
+        value: V,
+    }
+
+    /// The OrderedMap datastructure.
+    enum OrderedMap<K, V> has drop, copy, store {
+        /// sorted-vector based implementation of OrderedMap
+        SortedVectorMap {
+            /// List of entries, sorted by key.
+            entries: vector<Entry<K, V>>,
+        }
+    }
+
+    /// An iterator pointing to a position between two elements in the
+    enum Iterator has copy, drop {
+        // TODO: see if it is cleaner/faster to have end-iter be Position { index = len() }
+        // and not have a separate variant for it.
+        End,
+        Position {
+            /// The index of the iterator pointing to.
+            index: u64,
+        },
+    }
+
+    /// Creates a new empty OrderedMap, using default (SortedVectorMap) implementation.
+    public fun new<K, V>(): OrderedMap<K, V> {
+        OrderedMap::SortedVectorMap {
+            entries: vector::empty(),
+        }
+    }
+
+    /// Number of elements in the map.
+    public fun length<K, V>(self: &OrderedMap<K, V>): u64 {
+        self.entries.length()
+    }
+
+    /// Whether map is empty.
+    public fun is_empty<K, V>(self: &OrderedMap<K, V>): bool {
+        self.entries.is_empty()
+    }
+
+    /// Add a key/value pair to the map.
+    /// Aborts with EKEY_ALREADY_EXISTS if key already exist.
+    public fun add<K, V>(self: &mut OrderedMap<K, V>, key: K, value: V) {
+        let len = self.entries.length();
+        let index = binary_search(&key, &self.entries, 0, len);
+
+        // key must not already be inside.
+        assert!(index >= len || &self.entries[index].key != &key, error::invalid_argument(EKEY_ALREADY_EXISTS));
+        self.entries.insert(index, Entry { key, value });
+    }
+
+    /// If the key doesn't exist in the map, inserts the key/value, and returns none.
+    /// Otherwise, updates the value under the given key, and returns the old value.
+    public fun upsert<K: drop, V>(self: &mut OrderedMap<K, V>, key: K, value: V): Option<V> {
+        let len = self.entries.length();
+        let index = binary_search(&key, &self.entries, 0, len);
+
+        if (index < len && &self.entries[index].key == &key) {
+            let Entry {
+                key: _,
+                value: old_value,
+            } = self.entries.replace(index, Entry { key, value });
+            option::some(old_value)
+        } else {
+            self.entries.insert(index, Entry { key, value });
+            option::none()
+        }
+    }
+
+    /// Remove a key/value pair from the map.
+    /// Aborts with EKEY_NOT_FOUND if `key` doesn't exist.
+    public fun remove<K: drop, V>(self: &mut OrderedMap<K, V>, key: &K): V {
+        let len = self.entries.length();
+        let index = binary_search(key, &self.entries, 0, len);
+        assert!(index < len, error::invalid_argument(EKEY_NOT_FOUND));
+        let Entry { key: old_key, value } = self.entries.remove(index);
```

Comment:
vector does not keep the order here? IIRC, it `swap` and `pop_back`. 
This is critical.

Reply:
`remove` keeps the order. `swap_remove` doesn't